sentry-android-core/src/main/java/io/sentry/android/core/SpanFrameMetricsCollector.java

```diff
@@ -68,63 +107,130 @@ public void onSpanFinished(final @NotNull ISpan span) {
     if (!enabled) {
       return;
     }
+
     if (span instanceof NoOpSpan) {
       return;
     }
+
     if (span instanceof NoOpTransaction) {
       return;
     }
 
-    @Nullable SentryFrameMetrics diff = null;
+    // ignore span if onSpanStarted was never called for it
+    final @NotNull String spanId = span.getSpanContext().getSpanId().toString();
     synchronized (lock) {
-      final @Nullable SentryFrameMetrics metricsAtStart =
-          metricsAtSpanStart.remove(span.getSpanContext().getSpanId());
-      if (metricsAtStart != null) {
-        diff = currentFrameMetrics.diffTo(metricsAtStart);
+      if (!spanStarts.containsKey(spanId)) {
+        return;
+      }
+    }
+
+    captureFrameMetrics(span);
+
+    synchronized (lock) {
+      if (runningSpans.isEmpty()) {
+        clear();
+      } else {
+        // otherwise only remove old/irrelevant frames
+        final SpanStart oldestSpan = runningSpans.first();
+        frames.headSet(oldestSpan.timeNanos).clear();
       }
     }
+  }
 
-    if (diff != null && diff.containsValidData()) {
-      int nonRenderedFrameCount = 0;
+  private void captureFrameMetrics(@NotNull final ISpan span) {
+    // TODO lock still required?
+    synchronized (lock) {
+      final @NotNull String id = span.getSpanContext().getSpanId().toString();
+      final @Nullable NanoTimeStamp spanStartTimeStamp = spanStarts.remove(id);
+      if (spanStartTimeStamp == null) {
+        return;
+      }
+      runningSpans.remove(new SpanStart(id, spanStartTimeStamp));
 
-      // if there are no content changes on Android, also no frames are rendered
-      // thus no frame metrics are provided
-      // in order to match the span duration with the total frame count,
-      // we simply interpolate the total number of frames based on the span duration
-      // this way the data is more sound and we also match the output of the cocoa SDK
+      // ignore spans with no finish date
       final @Nullable SentryDate spanFinishDate = span.getFinishDate();
-      if (spanFinishDate != null) {
-        final long spanDurationNanos = spanFinishDate.diff(span.getStartDate());
+      if (spanFinishDate == null) {
+        return;
+      }
+
+      final @NotNull SentryFrameMetrics frameMetrics = new SentryFrameMetrics();
+
+      // project the span end date into our nano time format
+      final @NotNull SentryDate now = options.getDateProvider().now();
+      final long nowNanos = frameTimeProvider.now();
+
+      final long diffNanos = now.diff(spanFinishDate);
+      final long spanEndNanos = nowNanos - diffNanos;
+
+      final long spanStartNanos = spanStartTimeStamp.timeNanos;
+      if (spanStartNanos >= spanEndNanos) {
+        return;
+      }
+
+      final long spanDurationNanos = spanEndNanos - spanStartNanos;
```

Comment:
just trying to wrap my head around - why this wasn't working and why do we need a separate collection to track start timestamps?

```
 final long spanDurationNanos = spanFinishDate.diff(span.getStartDate());
```

Reply:
in this case it should work fine

Reply:
Because our `ISpans` only have `SentryDate`, which itself has no guarantees around precision - it could be e.g. a time synced unix timestamp. `SentryDate` is also not `Comparable`, so it can't be saved in an ordered set.
I'd be happy to change this within the whole SDK though. Maybe an extra one, to not blow this PR up too much 😅

Reply:
but `SentryDate` is an abstract class, in the end it'll be either `SentryNanotimeDate` or `SentryInstantDate` which both have the nanoseconds precision. And it's actually `Comparable` ;) https://github.com/getsentry/sentry-java/blob/a537f8a45828b751272940034c4a64ee5f8fd278/sentry/src/main/java/io/sentry/SentryDate.java#L48-L51

Reply:
Alright, I've looked into it and got it working [by applying a little trick](https://github.com/getsentry/sentry-java/blob/eedcf5ac53387d44c99f17cdc472ce0bb18dfb31/sentry-android-core/src/main/java/io/sentry/android/core/SpanFrameMetricsCollector.java#L298-L307)

Reply:
you can remove that trick if you use the `SentryNanotimeDate` in the `Frame` object, instead of nanoseconds